test/unit/eval_algorithms/test_context_quality.py

```diff
@@ -0,0 +1,138 @@
+        },
+        {
+            DatasetColumns.MODEL_INPUT.value.name: "What do the 3 dots mean in math?",
+            DatasetColumns.TARGET_OUTPUT.value.name: "therefore sign",
+            DatasetColumns.RETRIEVED_CONTEXT.value.name: "The therefore sign is generally used before a logical consequence. <AND>The symbol consists of three dots placed in an upright triangle and is read therefore.",
+        },
+    ]
+)
+
+
+class TestContextQuality:
+    @pytest.mark.parametrize(
+        "test_case",
+        [
+            TestCaseContextQualityEvaluateSample(
+                model_input="What is the tallest mountain in the world?",
+                target_output="Mount Everest.",
+                retrieved_context="The Andes is the longest continental mountain range in the world<AND> located in South America. <AND>It stretches across seven countries and features many of the highest peaks in the Western Hemisphere. <AND>The range is known for its diverse ecosystems, including the high-altitude Andean Plateau and the Amazon rainforest.",
+                mocked_judge_model_responses=[("The score is 0", None), ("0", None), ("1", None), ("unknown", None)],
+                expected_response=[EvalScore(name=CONTEXT_PRECISION_SCORE, value=0.3333333333)],
+            )
+        ],
+    )
+    def test_context_quality_evaluate_sample(self, test_case):
+        """
+        GIVEN a dataset sample with model_input, target_output and retrieved_context
+        WHEN evaluate_sample is called.
+        THEN the correct score is returned.
+        """
+        mock_runner = Mock()
+        mock_runner.predict.side_effect = test_case.mocked_judge_model_responses
+        config = ContextQualityConfig()
+
+        eval_algorithm = ContextQuality(config)
+        actual_response = eval_algorithm.evaluate_sample(
+            model_input=test_case.model_input,
+            target_output=test_case.target_output,
+            retrieved_context=test_case.retrieved_context,
+            judge_model=mock_runner,
+        )
+        assert actual_response == test_case.expected_response
+
+    @pytest.mark.parametrize(
+        "test_case",
+        [
+            TestCaseContextQualityEvaluate(
+                dataset=CONTEXT_QUALITY_DATASET,
+                data_config=DataConfig(
+                    dataset_name="retrieved_dataset",
+                    dataset_uri="path",
+                    dataset_mime_type=MIME_TYPE_JSONLINES,
+                    model_input_location="model_input",
+                    target_output_location="target_output",
+                    retrieved_context_location="retrieved_context",
+                ),
+                mocked_judge_model_responses=[
+                    ("0", None),
+                    ("The score is 1", None),
+                    ("1", None),
+                    ("It's hard to say", None),
+                    ("1", None),
```

Comment:
If my understanding is correct, the model should only get invoked 4 times, right?

Once for sample #1, once for sample #2 (since we should break out of the inner for loop early), and twice for sample #3 (because there are two contexts).